### PR TITLE
feat: adds implementation status to ssp-filter

### LINKED ISSUE #1332: Allow ssp-filter by implementation status
## Issue description / feature objectives
ssp-filter currently allows filtering based on controls (via profile) or components (via list of component names).  It could be useful to filter by implementation status - ideally a comma-separated list of status values
## Completion Criteria
Provide a new optional argument to `ssp-filter`:
`trestle author ssp-filter --implementation-status -is`
with value being a comma-separated list of status strings such as:
`implemented,partial`
Give error if one of the strings is not one of the allowed values for ssp implementation status.

### PR DESCRIPTION
**Pull request by jpower432**
## Types of changes

- [ ] Hot fix (emergency fix and release)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [X] My code follows the code style of this project.
- [X] Documentation for my change is up to date?
- [X] My PR meets testing requirements.
- [X] All new and existing tests passed.
- [X] All commits are signed-off.

## Summary

This PR adds an optional `implementation-status` flag to the `trestle author ssp-filter` command. This will filter implemented requirements and statements by the implementation status under `by_components`.

This PR also adds corresponding docs changes to update `ssp-filter` functionality.

Closes #1332 

## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

**Comment by fsuits** [comment]
@jpower432 this is looking very good - many thanks for offering to add this!  I have enabled running the workflow on push so it should work for you next time.  Unfortunately the workflows tend to die when running the windows tests for some reason in the cloud and not in the code - so if that happens please just restart any failed jobs.  When you are ready for review please add @enikonovad and me as reviewers and we will go through it.  Thanks!

**Comment by jpower432** [comment]
Thanks @fsuits. Happy to contribute! I will mark it as ready for review today. Just finishing up running some manual tests.

**Comment by jpower432** [comment]
@fsuits I don't have the permission to add reviewers, but this is ready for review. Thanks!

**Comment by fsuits** [comment]
@jpower432 Thanks!  And sorry the workflow was blocked.  I have assigned reviewers and will run the workflow.  Should be able to review it soon.

File: docs/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring.md [hunk 1]
```diff
@@ -1010,13 +1010,13 @@ If you do not specify component-defintions during assembly, the markdown should
...
 
-`trestle author ssp-filter --name my_ssp --profile my_profile --components comp_a:comp_b --output my_culled_ssp`
+`trestle author ssp-filter --name my_ssp --profile my_profile --components comp_a:comp_b --implementation-status "planned,partial" --output my_culled_ssp`
 
-Both the SSP and profile must be present in the trestle directory.  This command will generate a new SSP in the directory.  If the profile makes reference to a control not in the SSP then the routine will fail with an error message.  Similarly, if one of the components is not present in the ssp the routine will also fail.
+Both the SSP and profile must be present in the trestle directory.  This command will generate a new SSP in the directory.  If the profile makes reference to a control not in the SSP then the routine will fail with an error message.  Similarly, if one of the components is not present in the ssp the routine will also fail. The implementation statuses must be one of the allowed values as defined in the OSCAL SSP JSON format reference. Those include the following: implemented, partial, planned, alternative, and not-applicable. If an invalid value is provided, an error is returned.
```
Comment (fsuits) [inline docs/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring.md:1019]: The SSP must be present in the trestle workspace and, if filtering by profile, that profile must also be in the trestle workspace.  This command will generate a new SSP in the workspace.

File: trestle/core/commands/author/ssp.py [hunk 2]
```diff
@@ -580,20 +585,47 @@ def _init_arguments(self) -> None:
...
         try:
             log.set_log_level_from_args(args)
             trestle_root = pathlib.Path(args.trestle_root)
             comp_names: Optional[List[str]] = None
+            impl_status_values: Optional[List[str]] = None
+
+            if not (args.components or args.implementation_status or args.profile):
+                logger.warning(
+                    'You must specify either a profile, list of component names'
+                    ', or list of implementation statuses for ssp-filter.'
```
Comment (fsuits) [inline trestle/core/commands/author/ssp.py:601]: You must specify at least one, or a combination of: profile, ...

File: trestle/core/commands/author/ssp.py [hunk 3]
```diff
@@ -580,20 +585,47 @@ def _init_arguments(self) -> None:
...
             if args.components:
                 comp_names = args.components.split(':')
-            elif not args.profile:
-                logger.warning('You must specify either a profile or list of component names for ssp-filter.')
-                return 1
+
+            if args.implementation_status:
+                impl_status_values = args.implementation_status.split(',')
+                allowed_is_values = {'planned', 'partial', 'implemented', 'alternative', 'not-applicable'}
+                allowed_is_string = ', '.join(str(item) for item in allowed_is_values)
```
Comment (fsuits) [inline trestle/core/commands/author/ssp.py:611]: Please use corresponding constants defined in const.py, e.g. `STATUS_PLANNED`

File: trestle/core/commands/author/ssp.py [hunk 4]
```diff
@@ -706,6 +743,35 @@ def filter_ssp(
 
             ssp.control_implementation = control_imp
 
+        # filter implemented requirements and statements by component implementation status
+        # this will remove any implemented requirements without statements or by_component fields set
+        if implementation_status:
+            new_imp_reqs: List[ossp.ImplementedRequirement] = []
+            # these are all required to be present
+            for imp_req in ssp.control_implementation.implemented_requirements:
+                new_by_comps: List[ossp.ByComponent] = []
```
Comment (fsuits) [inline trestle/core/commands/author/ssp.py:752]: use `as_list(ssp.control...)` because imp_reqs may be None

**Review by fsuits — APPROVED** [review]
LGTM Looks good.  Please address small doc and style changes.

File: trestle/core/commands/author/ssp.py [hunk 5]
```diff
@@ -580,20 +585,47 @@ def _init_arguments(self) -> None:
...
+                impl_status_values = args.implementation_status.split(',')
+                allowed_is_values = {'planned', 'partial', 'implemented', 'alternative', 'not-applicable'}
+                allowed_is_string = ', '.join(str(item) for item in allowed_is_values)
+                for impl_status in impl_status_values:
+                    if impl_status not in allowed_is_values:
+                        logger.warning(
+                            f'Provided implementation status "{impl_status}" is invalid.\n'
+                            f'Please use the following for ssp-filter: {allowed_is_string}'
+                        )
+                        return 1
```
Comment (enikonovad) [inline trestle/core/commands/author/ssp.py:618]: Please use `CmdReturnCodes.COMMAND_ERROR.value` here instead of '1'

**Review by enikonovad — APPROVED** [review]
LGTM - after the small changes above

**Review by fsuits — APPROVED** [review]
LGTM

**Pull request merged**